### PR TITLE
Project initialization

### LINKED ISSUE #4: Project initialization
- [x] node.js project initialization
- [x] Adding TypeScript
- [x] Adding linters

### PR DESCRIPTION
**Pull request by vketrys**
- [x] Create node.js TypeScript app
- [x] Configuration
- [x] Adding linters
- [x] Connect Firebase
- [x] Creating SignUp/SignIn logic

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,7 +5,8 @@
   "main": "index.js",
   "type": "module",
   "scripts": {
-    "build": "tsc",
+    "build": "tsc -w",
+    "dev": "nodemon ./dist/index.js",
```
Comment (esnts) [inline package.json]: why we are using index.js instead of .ts file?

File: package.json [hunk 2]
```diff
@@ -21,11 +22,19 @@
   },
   "homepage": "https://github.com/vketrys/nodejs-app#readme",
   "devDependencies": {
+    "@types/express": "^4.17.14",
```
Comment (esnts) [inline package.json]: remove version autoincrementation

File: src/auth.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import { 
+  getAuth, 
+  createUserWithEmailAndPassword, 
+  signInWithEmailAndPassword
```
Comment (esnts) [inline src/auth.ts]: LINT?

File: src/auth.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import { 
+  getAuth, 
+  createUserWithEmailAndPassword, 
+  signInWithEmailAndPassword 
+} from 'firebase/auth';
+import app from './config/firebase.js';
+
+export const signup = (req, res) => {
```
Comment (esnts) [inline src/auth.ts]: signup type? 
req type? 
res type?

Comment (esnts) [reply src/auth.ts]: export const signup = ({body : {.....}}, res) => {

File: src/auth.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  if (!req.body.email || !req.body.password) {
+    return res.status(422).json({
```
Comment (esnts) [inline src/auth.ts]: 422 -> constants

File: src/auth.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      email: 'email is required',
```
Comment (esnts) [inline src/auth.ts]: `email is required` to C
...

File: src/auth.ts (hunk 4 above)
Comment (esnts) [inline src/auth.ts]: try catch finally(optional)

File: src/auth.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+import { 
+  getAuth, 
+  createUserWithEmailAndPassword, 
+  signInWithEmailAndPassword 
+} from 'firebase/auth';
+import app from './config/firebase.js';
```
Comment (esnts) [inline src/auth.ts:7]: WHY JS? WTFFFFF???

File: src/auth.ts [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  if (!req.body.email || !req.body.password) {
+    return res.status(422).json({
+      email: 'email is required',
+      password: 'password is required',
+    });
+  }
+  const auth = getAuth(app);
+  createUserWithEmailAndPassword(auth, req.body.email, req.body.password)
+    .then((data) => {
+      return res.status(201).json(data);
```
Comment (esnts) [inline src/auth.ts]: 201 - C

File: src/auth.ts [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    })
+    .catch(function (error) {
+      if (error.code == 'auth/weak-password') {
```
Comment (esnts) [inline src/auth.ts]: Can you get this S from node_module ? 'auth/weak-password'

Comment (esnts) [reply src/auth.ts]: or -> C

Comment (esnts) [reply src/auth.ts]: ===

Comment (esnts) [reply src/auth.ts]: return res.status(500).json({ error: error.message });

File: src/index.ts [hunk 10]
```diff
@@ -1,2 +1,14 @@
...
 
+const app = express();
+const port = 3000;
+
+app.use(bodyParser.json());
+
+app.use("/api", authRoutes);
+
+app.listen(port, () => {
+  console.log(`app listening on port ${port}`);
```
Comment (esnts) [inline src/index.ts]: NO LOGS

Comment (esnts) [reply src/index.ts]: + log to output file

File: src/index.ts [hunk 11]
```diff
@@ -1,2 +1,14 @@
...
+import express from 'express';
+import authRoutes from './routes/routes.js'
+import bodyParser from 'body-parser';
 
+const app = express();
+const port = 3000;
+
+app.use(bodyParser.json());
+
+app.use("/api", authRoutes);
```
Comment (esnts) [inline src/index.ts]: -> C

File: src/routes/routes.ts [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import express from 'express';
+import { signup, signin } from '../auth.js'
+
+const router = express.Router();
+
+router.post("/signup", signup);
```
Comment (esnts) [inline src/routes/routes.ts]: /signup -> C 

const URL = {
ROOT : "/api",
AUT: {
SIGNIN: '/signin'
} 
}

**Review by esnts — CHANGES_REQUESTED** [review]

File: src/auth.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+	signInWithEmailAndPassword,
+} from 'firebase/auth';
+import { Request, Response } from 'express';
+import app from './config/firebase.js';
+import { responses } from './constants/responses.js';
+import { errorCodes, statusCodes } from './constants/codes.js';
+
+export const signup = async(req: Request, res: Response): Promise<Response> => {
+	try {
+		const { email, password } = req.body;
```
Comment (esnts) [inline src/auth.ts:13]: tab 8 symbols? super strange

File: src/auth.ts [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+  
+		if (!email || !password) {
+			return res.status(statusCodes.unprocessableEntity).json({
```
Comment (esnts) [inline src/auth.ts:16]: never send passwords

File: src/auth.ts [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+				email: responses.emailRequired,
+				password: responses.passwordRequired,
+			});
+		}
+  
+		const auth = getAuth(app);
+  
+		createUserWithEmailAndPassword(auth, email, password)
+			.then((user) => {
```
Comment (esnts) [inline src/auth.ts:25]: попробуй найти правило для линта. не оборачивать функцию в скобки если там только ретурн

File: src/auth.ts [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+				return res.status(statusCodes.created).json(user);
+			});
+	} catch (error) {
+		if (error.code === errorCodes.weakPassword) {
```
Comment (esnts) [inline src/auth.ts:29]: res.status(error.code === errorCodes.weakPassword ? statusCodes.badRequest : .... )
   .json({ error: error.message });
   
or

create const for error.code === errorCodes.weakPassword ? statusCodes.badRequest :

File: src/auth.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+			return res.status(statusCodes.unprocessableEntity).json({
+				email: responses.emailRequired,
+				password: responses.passwordRequired,
+			});
+		}
+  
+		const auth = getAuth(app);
+  
+		signInWithEmailAndPassword(auth, email, password)
+			.then((user) => {
```
Comment (esnts) [inline src/auth.ts:51]: Какой тип юзера, мы точно не отправляем лишней информации?

File: src/config/firebase.ts [hunk 18]
```diff
@@ -0,0 +1,17 @@
+import { initializeApp } from 'firebase/app';
+import dotenv from 'dotenv';
+dotenv.config();
+
+const firebaseConfig = {
+	apiKey: process.env.API_KEY,
```
Comment (esnts) [inline src/config/firebase.ts:6]: NICE

File: src/constants/URL.ts [hunk 19]
```diff
@@ -0,0 +1,7 @@
+export const URL = {
+	ROOT: '/api',
+	AUTH: {
+		SIGNIN: '/signin',
```
Comment (esnts) [inline src/constants/URL.ts:4]: nice

File: src/index.ts [hunk 20]
```diff
@@ -0,0 +1,16 @@
+import express, { Express } from 'express';
+import authRoutes from './routes/routes.js';
+import bodyParser from 'body-parser';
+import { URL } from './constants/URL.js';
+
+const app: Express = express();
+const port = process.env.PORT || 3000;
```
Comment (esnts) [inline src/index.ts:7]: nice

**Review by esnts — APPROVED** [review]
LGTM

**Review by slamser56 — APPROVED** [review]

**Pull request merged**